mex/common/types/timestamp.py

```diff
@@ -250,3 +251,84 @@ def __str__(self) -> str:
         return f'{self.__class__.__name__}("{self}")'
+
+
+class TimestampYearMonth(Timestamp):
+    """Partial date pattern that accepts yyyy or yyyy-MM format."""
+
+    def __init__(self, *args: int, tzinfo: Optional[tzinfo] = None) -> None:
+        """Validate for precision (year or month) after initialization."""
+        super().__init__(*args, tzinfo=tzinfo)
+        if self.precision not in (TimestampPrecision.YEAR, TimestampPrecision.MONTH):
+            raise ValueError("Expected precision level 'YEAR' or 'MONTH'")
+
+    @classmethod
+    def __get_pydantic_core_schema__(
+        cls, _source: Type[Any], _handler: GetCoreSchemaHandler
+    ) -> core_schema.CoreSchema:
+        """Mutate the field schema for year-month pattern."""
+        from_str_schema = core_schema.chain_schema(
+            [
+                core_schema.str_schema(pattern=YEAR_MONTH_REGEX),
+            ]
+        )
+
+        from_anything_schema = core_schema.chain_schema(
+            [
+                core_schema.no_info_plain_validator_function(cls.validate),
+            ]
+        )
+
+        return core_schema.json_or_python_schema(
+            json_schema=from_str_schema,
+            python_schema=from_anything_schema,
+        )
+
+    @classmethod
+    def __get_pydantic_json_schema__(
+        cls, core_schema_: CoreSchema, handler: GetJsonSchemaHandler
+    ) -> JsonSchemaValue:
+        """Modify the schema to add the class name as title and examples."""
+        json_schema = super().__get_pydantic_json_schema__(core_schema_, handler)
+        json_schema["examples"] = ["2011", "2019-03"]
+        return json_schema
+
+
+class TimestampDate(Timestamp):
+    """Date pattern that accepts only yyyy-MM-dd format."""
+
+    def __init__(self, *args: int, tzinfo: Optional[tzinfo] = None) -> None:
+        """Validate for day precision after initialization."""
+        super().__init__(*args, tzinfo=tzinfo)
+        if self.precision != TimestampPrecision.DAY:
+            raise ValueError("Expected precision level 'DAY'")
+
+    @classmethod
+    def __get_pydantic_json_schema__(
+        cls, core_schema_: CoreSchema, handler: GetJsonSchemaHandler
+    ) -> JsonSchemaValue:
+        """Modify the schema to add the class name as title and examples."""
+        json_schema = super().__get_pydantic_json_schema__(core_schema_, handler)
+        json_schema["examples"] = ["2011-03-21"]
+        json_schema["format"] = "date"
+        return json_schema
+
+
+class TimestampDatetime(Timestamp):
+    """Date pattern that accepts only yyyy-MM-dd HH:mm:ss format."""
+
+    def __init__(self, *args: int, tzinfo: Optional[tzinfo] = None) -> None:
+        """Validate for timestamp (up to seconds) precision after initialization."""
+        super().__init__(*args, tzinfo=tzinfo)
+        if self.precision != TimestampPrecision.SECOND:
+            raise ValueError("Expected precision level 'SECOND'")
```

Comment:
```suggestion
        if self.precision not in TIME_PRECISIONS:
            raise ValueError("Expected time-based precision level")
```

i think we could/should allow the hour/minute/ms here too

Reply:
hmm but like this we allow all precisions, even year etc.?

Reply:
should just be these 4, or am i missing sth?

```python
TIME_PRECISIONS = [
    TimestampPrecision.HOUR,
    TimestampPrecision.MINUTE,
    TimestampPrecision.SECOND,
    TimestampPrecision.MICROSECOND,
]
```

Reply:
you're absolutely right, thought it includes all precisions, will change it!